addon/mixins/callback-route-mixin.js

```diff
@@ -0,0 +1,69 @@
+import Ember from 'ember';
+
+const { inject: { service }, Mixin } = Ember;
+
+export default Mixin.create({
+  /**
+   The session service.
+
+   @property session
+   @readOnly
+   @type SessionService
+   @public
+   */
+  session: service('session'),
+
+  /**
+   The authenticator that should be used to authenticate the callback.
+
+   @property authenticator
+   @type String
+   @default null
+   @public
+   */
+  authenticator: null,
+
+  /**
+   The route that should be transitioned to on error.
+
+   @property routeOnError
+   @type String
+   @default null
+   @public
+   */
+  routeOnError: null,
+
+  /**
+   The route that should be transitioned to on success.
+
+   @property routeOnSuccess
+   @type String
+   @default null
+   @public
+   */
+  routeOnSuccess: null,
+
+  /**
+   Authenticates the passed authentication server response and transitions
+   to the `routeOnError` route with the error specified as query parameters
+   or to the `routeOnSuccess`route.
+
+   If the current windows is a pop-up it will be automatically closed on success.
+
+   @method activate
+   @public
+   */
+  activate() {
+    let authenticator = this.get('authenticator');
+
+    this.get('session').authenticate(authenticator).then(() => {
+      if (window.opener) {
+        return window.close();
+      }
+
+      return this.transitionTo(this.get('routeOnSuccess'));
+    }).catch((err) => {
+      return this.transitionTo(this.get('routeOnError'), { queryParams: err });
+    });
```

Comment:
I'll remove the unnecessary success transition, but in most cases the user wants to transition back to the route that initiated the authentication process. This is the use case this mixin should cover. I'll make the error transition optional and also set the error in a property to cover both cases.